controllers/manilascheduler_controller.go

```diff
@@ -323,25 +323,44 @@ func (r *ManilaSchedulerReconciler) reconcileNormal(ctx context.Context, instanc
 	//
 	// check for required OpenStack secret holding passwords for service/admin user and add hash to the vars map
 	//
-	ctrlResult, err := r.getSecret(ctx, helper, instance, instance.Spec.Secret, &configVars)
+
+	hash, ctrlResult, err := ensureSecret(
```

Comment:
I checked the code, and the subCRs do not use `instance.Spec.Secret` other than setting config options referencing the string specified in the CR.
+1 to remove this extra call.

Reply:
I think I need to keep this call here and I'll try to explain the reason.
It's true that the top-level controller looks for the existence of `osp-secret` before creating the subCRs, and we're able to block the reconciliation there.
However, when the secret exists and the subCRs (manilaAPI, manilaScheduler, manilaShare) are created, the control moves to the sub-level controller (each resource has a dedicated controller). 
What I do here is ensuring the secret exists, and saving the hash in the status. If the `osp-secret` is updated (because the manila password changes), the goal is to rollout new Pods because the input changed, and ignoring any update to fields not related to `manila`.
Without this check, any subsequent reconciliation loop triggered for the subCR will ignore an update to the secret. For this reason I need to keep this call here.

Reply:
> I agree we need to keep using ensureSecret(), but I reached the conclusion for different reasons. I'm not sure whether the fact that the services have their own controller matters, but what does matter is they need to leverage ensureSecret()'s ability to return the hash for just the service's own password, and not the entire secret.

Correct, and that hash is what triggers a potential rollout of new Pods is the secret changes. I started from the reconciliation aspect in the previous comment and I was trying to describe the flow, but in the end the hash is what matters for future reconciliation loops.
> 
> With this in mind, since I'm working on the corresponding cinder-operator change, what do you think of renaming it to getServiceSecret() and letting it deal with the expectedFields[] stuff? That might reduce the amount of code that's repeated in the sub-services.

I'm ok to rename `ensureSecret` to `getServiceSecret`, especially if it helps to keep consistency across the storage operators. I'm going to update this change with a couple of more optimizations.
> 
> I expect to have my patch ready tomorrow. It might be worth holding of merging either patch until we iron out all the details, so that we can cross pollinate our ideas.

I'll put this one on /hold, so we can discuss more and converge into a shared idea. Unfortunately I'll be on PTO both Thu and Fri, so we might need to wait until Monday to sync up. Feel free to move forward with your work though, and then we'll discuss more according to your findings.